File: python/ai_orchestrator/schemas/assistant_request.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


LearningMode = Literal["daily_explain", "exam_boost"]
AssistantIntent = Literal[
    "free_chat",
    "explain",
    "translate",
    "polish",
    "summarize",
    "grade_writing",
    "generate_examples",
    "analyze_question",
]
InputScope = Literal[
    "message_only",
    "selection",
    "attachments",
    "selection_and_message",
    "attachments_and_message",
    "selection_attachments_and_message",
]
SelectionSource = Literal[
    "assistant_message",
    "writing_editor",
    "page_selection",
    "uploaded_image_ocr",
]
AttachmentProvider = Literal["app_storage", "openai_files", "external_url"]
AttachmentKind = Literal["image", "pdf", "txt", "docx", "doc", "other"]
AttachmentProcessingStatus = Literal["uploaded", "processing", "ready", "failed"]
PreferredModelInputPart = Literal["input_image", "input_file", "input_text"]
ImageDetail = Literal["low", "high", "auto"]


class AssistantAttachmentProcessing(BaseModel):
    status: AttachmentProcessingStatus
    error_code: str | None = Field(default=None, alias="errorCode")
    extracted_text_available: bool | None = Field(default=None, alias="extractedTextAvailable")
    extracted_text: str | None = Field(default=None, alias="extractedText")
    page_count: int | None = Field(default=None, alias="pageCount")
    checksum: str | None = None

    model_config = {"populate_by_name": True}


class AssistantAttachmentModelInput(BaseModel):
    preferred_part: PreferredModelInputPart | None = Field(default=None, alias="preferredPart")
    image_detail: ImageDetail | None = Field(default=None, alias="imageDetail")

    model_config = {"populate_by_name": True}


class AssistantAttachmentRef(BaseModel):
    attachment_id: str = Field(alias="attachmentId")
    provider: AttachmentProvider
    openai_file_id: str | None = Field(default=None, alias="openaiFileId")
    storage_key: str | None = Field(default=None, alias="storageKey")
    url: str | None = None
    name: str
    mime_type: str = Field(alias="mimeType")
    size_bytes: int = Field(alias="sizeBytes")
    kind: AttachmentKind
    processing: AssistantAttachmentProcessing
    model_input: AssistantAttachmentModelInput | None = Field(default=None, alias="modelInput")

    model_config = {"populate_by_name": True}


class AssistantSelectionRange(BaseModel):
    start: int | None = None
    end: int | None = None


class AssistantSelection(BaseModel):
    text: str
    source: SelectionSource
    source_id: str | None = Field(default=None, alias="sourceId")
    message_id: str | None = Field(default=None, alias="messageId")
    document_id: str | None = Field(default=None, alias="documentId")
    range: AssistantSelectionRange | None = None

    model_config = {"populate_by_name": True}


class AssistantRequestMessage(BaseModel):
    text: str | None = None


class AssistantStudyContext(BaseModel):
    study_stage: Literal["beginner", "intermediate", "advanced"] | None = Field(default=None, alias="studyStage")
    cefr_level: Literal["A1", "A2", "B1", "B2", "C1", "C2"] | None = Field(default=None, alias="cefrLevel")
    target_exam: Literal["ielts", "toefl", "cet4", "cet6", "gaokao"] | None = Field(default=None, alias="targetExam")
    locale: Literal["zh-CN", "en-US"] | None = None
    response_language: Literal["zh-CN", "en-US", "mixed"] | None = Field(default=None, alias="responseLanguage")

    model_config = {"populate_by_name": True}


class AssistantClientMeta(BaseModel):
    source_page: str | None = Field(default=None, alias="sourcePage")
    timezone: str | None = None
    user_agent: str | None = Field(default=None, alias="userAgent")

    model_config = {"populate_by_name": True}


class AssistantRequest(BaseModel):
    app_conversation_id: str | None = Field(default=None, alias="appConversationId")
    client_message_id: str = Field(alias="clientMessageId")
    idempotency_key: str | None = Field(default=None, alias="idempotencyKey")
    mode: LearningMode
    intent: AssistantIntent
    scope: InputScope | None = None
    message: AssistantRequestMessage = Field(default_factory=AssistantRequestMessage)
    selection: AssistantSelection | None = None
    attachments: list[AssistantAttachmentRef] = Field(default_factory=list)
    study_context: AssistantStudyContext | None = Field(default=None, alias="studyContext")
    client_meta: AssistantClientMeta | None = Field(default=None, alias="clientMeta")

    model_config = {"populate_by_name": True}


class AssistantUsage(BaseModel):
    input_tokens: int | None = Field(default=None, alias="inputTokens")
    output_tokens: int | None = Field(default=None, alias="outputTokens")
    total_tokens: int | None = Field(default=None, alias="totalTokens")
    requests: int | None = None

    model_config = {"populate_by_name": True}


class AssistantOpenAIState(BaseModel):
    response_id: str | None = Field(default=None, alias="responseId")
    conversation_id: str | None = Field(default=None, alias="conversationId")
    previous_response_id: str | None = Field(default=None, alias="previousResponseId")

    model_config = {"populate_by_name": True}


class AssistantRunMetadata(BaseModel):
    run_id: str = Field(alias="runId")
    trace_id: str | None = Field(default=None, alias="traceId")
    agent_name: str = Field(alias="agentName")
    model: str
    mode: LearningMode
    intent: AssistantIntent
    scope: InputScope
    finish_reason: str | None = Field(default=None, alias="finishReason")

    model_config = {"populate_by_name": True}

File: web/src/types/assistantRequest.ts
import type { AssistantBlock } from './assistantBlocks.ts'

export type LearningMode = 'daily_explain' | 'exam_boost'

export type AssistantInteractionSource =
  | 'composer'
  | 'quick_action'
  | 'response_action'
  | 'activity_action'

export interface AssistantInteractionContext {
  source: AssistantInteractionSource
  uiIntent?: 'start_practice' | 'show_learning_card' | 'activity_action'
  activeActivityId?: string
  actionId?: string
  context?: {
    exerciseType?: 'sentence_reorder'
    topic?: string
    difficulty?: 'easy' | 'medium' | 'hard'
  }
}

export interface AssistantInteractionTrigger {
  displayText: string
  interaction: AssistantInteractionContext
}

export type AssistantIntent =
  | 'free_chat'
  | 'explain'
  | 'translate'
  | 'polish'
  | 'summarize'
  | 'grade_writing'
  | 'first_draft_coach'
  | 'generate_examples'
  | 'analyze_question'

export type InputScope =
  | 'message_only'
  | 'selection'
  | 'attachments'
  | 'selection_and_message'
  | 'attachments_and_message'
  | 'selection_attachments_and_message'

export type SelectionSource =
  | 'assistant_message'
  | 'writing_editor'
  | 'page_selection'
  | 'uploaded_image_ocr'

export type AttachmentProvider = 'app_storage' | 'openai_files' | 'external_url'

export type AttachmentKind = 'image' | 'pdf' | 'txt' | 'docx' | 'doc' | 'other'

export type AttachmentProcessingStatus = 'uploaded' | 'processing' | 'ready' | 'failed'

export type PreferredModelInputPart = 'input_image' | 'input_file' | 'input_text'

export type ImageDetail = 'low' | 'high' | 'auto'

export type ResponseLanguage = 'zh-CN' | 'en-US' | 'mixed'

export type StudyStage = string

export type CefrLevel = 'A1' | 'A2' | 'B1' | 'B2' | 'C1' | 'C2'

export type TargetExam = 'ielts' | 'toefl' | 'cet4' | 'cet6' | 'gaokao' | 'postgrad'

export interface AssistantAttachmentRef {
  attachmentId: string
  provider: AttachmentProvider
  openaiFileId?: string
  storageKey?: string
  url?: string
  name: string
  mimeType: string
  sizeBytes: number
  kind: AttachmentKind
  processing: {
    status: AttachmentProcessingStatus
    errorCode?: string
    extractedTextAvailable?: boolean
    extractedText?: string
    pageCount?: number
    checksum?: string
  }
  modelInput?: {
    preferredPart?: PreferredModelInputPart
    imageDetail?: ImageDetail
  }
}

export interface AssistantSelection {
  text: string
  source: SelectionSource
  sourceId?: string
  messageId?: string
  documentId?: string
  range?: {
    start?: number
    end?: number
  }
}

export interface AssistantConversationHistoryMessage {
  role: 'user' | 'assistant'
  content: string
}

export interface AssistantWritingCoachContext {
  schemaVersion?: string
  action?: 'coach' | 'analyze' | 'outline' | 'next' | 'topic' | 'polish' | 'draft'
  writingMode?: 'free' | 'exam'
  studyStage?: string | null
  taskType?: string | null
  essayQuestion?: string | null
  questionMaterials?: string | null
  imageDescriptions?: string[]
  attachments?: AssistantAttachmentRef[]
  essayGenre?: string | null
  minWords?: number | null
  maxWords?: number | null
  draftText?: string | null
  selectedText?: string | null
  includeDraft?: boolean
  topicAnalysisDone?: boolean
  topicBrief?: string | null
  centralTask?: string | null
  mustAnswerPoints?: string[]
  riskPoints?: string[]
  recommendedStructure?: string[]
  rubric?: {
    rubricKey?: string
    rubricVersion?: string
    rubricText?: string
    rubricFocus?: string[]
  }
}

export type WritingCoachEditActionType =
  | 'replace_selection'
  | 'insert_after_selection'
  | 'append_paragraph'

export type WritingPatch =
  | {
      op: 'replace_selection'
      range: { start: number; end: number }
      originalText: string
      newText: string
      reason?: string
    }
  | {
      op: 'search_replace'
      searchText: string
      replaceText: string
      reason?: string
    }
  | {
      op: 'insert_after_anchor'
      anchorText: string
      insertText: string
      reason?: string
    }
  | {
      op: 'append_paragraph'
      text: string
      reason?: string
    }
  | {
      op: 'replace_document'
      text: string
      reason?: string
    }

export interface WritingCoachEditAction {
  id: string
  type: WritingCoachEditActionType
  title: string
  text: string
  reason?: string
  patch?: WritingPatch
  target?: {
    mode: 'selected_range' | 'semantic_match' | 'document_end'
    selectedText?: string
    range?: {
      start: number
      end: number
    }
  }
}

export interface AssistantRequest {
  appConversationId?: string
  clientMessageId: string
  idempotencyKey?: string
  mode: LearningMode
  intent: AssistantIntent
  scope?: InputScope
  message: {
    text?: string
  }
  interaction?: AssistantInteractionContext
  selection?: AssistantSelection
  attachments?: AssistantAttachmentRef[]
  studyContext?: {
    studyStage?: StudyStage
    cefrLevel?: CefrLevel
    targetExam?: TargetExam
    locale?: 'zh-CN' | 'en-US'
    responseLanguage?: ResponseLanguage
  }
  writingCoachContext?: AssistantWritingCoachContext
  clientMeta?: {
    sourcePage?: string
    timezone?: string
    userAgent?: string
  }
  conversationHistory?: AssistantConversationHistoryMessage[]
}

export interface AssistantRunMetadata {
  runId: string
  traceId?: string
  agentName: string
  model: string
  mode: LearningMode
  intent: AssistantIntent
  scope: InputScope
  finishReason?: string
}

export interface AssistantUsage {
  inputTokens?: number
  outputTokens?: number
  totalTokens?: number
  requests?: number
}

export interface AssistantOpenAiState {
  responseId?: string
  conversationId?: string
  previousResponseId?: string
}

export interface AssistantMessageResponse {
  appConversationId: string
  messageId: string
  role: 'assistant'
  content: string
  parts?: AssistantBlock[]
  run: AssistantRunMetadata
  usage?: AssistantUsage
  openai?: AssistantOpenAiState
  createdAt: string
}

export interface AssistantErrorPayload {
  code:
    | 'INVALID_REQUEST'
    | 'IDEMPOTENCY_CONFLICT'
    | 'MISSING_INPUT'
    | 'UNSUPPORTED_INTENT'
    | 'UNSUPPORTED_MODE'
    | 'ATTACHMENT_NOT_READY'
    | 'ATTACHMENT_TOO_LARGE'
    | 'ATTACHMENT_KIND_UNSUPPORTED'
    | 'ATTACHMENT_IMAGE_NOT_READABLE'
    | 'ATTACHMENT_FILE_NOT_READABLE'
    | 'MODEL_CAPABILITY_UNSUPPORTED'
    | 'GUARDRAIL_BLOCKED'
    | 'OPENAI_RUN_FAILED'
    | 'STREAM_CANCELLED'
    | 'TIMEOUT'
  message: string
  details?: unknown
}

export type AssistantStreamEvent =
  | {
      type: 'run.started'
      runId: string
      traceId?: string
      agentName: string
      model: string
    }
  | {
      type: 'handoff'
      runId: string
      fromAgent: string
      toAgent: string
    }
  | {
      type: 'message.created'
      runId: string
      messageId: string
      role: 'assistant'
    }
  | {
      type: 'message.delta'
      runId: string
      messageId: string
      delta: string
    }
  | {
      type: 'message.completed'
      runId: string
      messageId: string
      content: string
      parts?: AssistantBlock[]
    }
  | {
      type: 'run.completed'
      runId: string
      usage?: AssistantUsage
      openai?: AssistantOpenAiState
    }
  | {
      type: 'run.failed'
      runId: string
      error: AssistantErrorPayload
    }
